### DISCUSSION TITLE
About a VoxCeleb2 helper script

### DISCUSSION THREAD

**Discussion by underdogliu** [speechbrain/speechbrain#1647] [Q&A]
Hi all,

So in VoxCeleb recipe users are required to create wav files by themselves. I personally am perfectly OK with this as a speaker recognition researcher (have done it many times) but meanwhile think having a helper recipe/function would be very helpful. 

I do have scripts available and I would like to contribute. But it has two versions so I would like to know the dev guys' opinion:
1. We process the m4a file one-by-one. This does not require any dependency but takes time (I don't think several hours would be enough).
2. We split the IDs into different processes. But I think this process cannot be done efficiently because of GIL. Of course, if we have tools like [Kaldi](https://github.com/kaldi-asr/kaldi), it would not be a issue because Kaldi can take advantages of multi-core by its perl runners. But of course, we need Kaldi somewhere, which I am not sure if fits the logistics of this toolkit.

What is your view on this? Do you have any alternative solutions? If so I would appreciate it. Thanks in advance!

**Comment by mravanelli** [comment]
Hi, this is rather annoying. The main issue is that torchaudio cannot read
m4a and thus users have to do the conversion to a readable format (e.g,
wav). This is discussed also in the torchaudio github (
https://github.com/pytorch/audio/issues/104). On our side, we point in the
README file to some bash scripts that can be used to do the conversion.

I think the ideal solution would be to do everything with a python library
within voxceleb_prepare.py. I'm not sure if there are easy-to-install
libraries that can manage that though. To speed it up we can consider
multi-processing as well (even though donging in python is not ideal).  As
for the time, however, I'm not that concerned. It is likely that users will
do it only once and then they save the converted version somewhere in their
system (as I did). Feel free to propose a solution that you think can work
well!

**Comment by underdogliu** [comment]
@mravanelli Thanks for the reply! So the problem is actually on lacking tools from torchaudio on reading m4a file to raw waveform. Am I right? If so, as mentioned in the issue, pydub may do the job. But maybe we want to have torchaudio as the only library for audio processing?

**Comment by mthrok** [comment]
(Acknowledged by torchaudio team.)

**Comment by mravanelli** [comment]
Yes

**Comment by underdogliu** [comment]
One direct solution is to integrate the "m4a->wav" process into the code, as streaming stuff. This can be done very effectively by calling `os.system($cmd)`, where `$cmd` include `ffmpeg` operations.

Another solution is to either wait torchaudio dev genius or we finding a way to integrate some way of reading m4a files into the library. But in the case of this solution, I am not sure if we want to rely on particular third-party libs like pydub then - maybe we can get inspired.

@mthrok @mravanelli thoughts?

**Comment by BenoitWang** [comment]
Hi, thanks @underdogliu for the question and I found this inconvenience too when I tried to run a benchmark with voxceleb1+2. I am running ffmpeg as mentioned in readme while waiting for the response of torchaudio team.

**Comment by mravanelli** [comment]
The cleanest solution would be to support m4a reading from torchaudio. This
way, other projects can benefit from it.
We can potentially integrate the call to ffmpeg using os.system($cmd).This
is a solution that we wanted to avoid because it makes the code dependent
on the operating system (and the tools installed in it).

**Comment by mravanelli** [comment]
@mthrok is there any plan in torchaudio to support m4a format?

**Comment by mthrok** [comment]
It has been added to the future plan and I bumped up the priority. We should be able to work on this in the next quarter. (It is not a guarantee though)

**Comment by underdogliu** [comment]
@mthrok anything I can help?

**Comment by underdogliu** [comment]
@mravanelli So just a quick coming back on this issue. One option is definitely calling back Kaldi. Torchaudio by default supports [kaldi-io](https://pytorch.org/audio/stable/kaldi_io.html) as we all know. And for reading voxceleb's m4a files with necessary utilities (ffmpeg), we can split and load from Kaldi scp files. But I remember one kinda "selling point" of speechbrain is getting away from Kaldi. So what are your thoughts on this?

**Comment by mravanelli** [comment]
I had a discussion with torchaudio. They will likely support m4a files in
one of the future versions. Once that is done, it will be super easy to
read these files without any intermediate conversion to wav.
